### DISCUSSION TITLE
Problem loading custom model

### DISCUSSION THREAD

**Discussion by Corallo** [AxisCommunications/acap-computer-vision-sdk-examples#62] [Q&A] [answered]
Hi, I have the same issue when I loaded my own tflite model. 
I have already replaced the path to my model in the [Dockerfile.model](https://github.com/AxisCommunications/acap-computer-vision-sdk-examples/blob/main/object-detector-python/Dockerfile.model) and [env.aarch64.artpec8](https://github.com/AxisCommunications/acap-computer-vision-sdk-examples/blob/main/object-detector-python/config/env.aarch64.artpec8):

Dockerfile.model:
```
# Get custom model
ADD ./models/yolov4/train-608_best.tflite models/
ADD ./models/yolov4/labels.txt models/
```
env.aarch64.artpec8:
```
MODEL_PATH=/models/train-608_best.tflite
OBJECT_LIST_PATH=/models/labels.txt
INFERENCE_SERVER_IMAGE=axisecp/acap-runtime:aarch64-containerized
INFERENCE_SERVER_ARGS="-m /models/train-608_best.tflite -j 12" 
```

There is the error I encountered:
![image](https://user-images.githubusercontent.com/59949085/182321422-ded4e6e2-915c-4eb6-9589-e12764be0384.png)

error message:

>  ERROR in Inference: Failed to load model train-608_best.tflite (Could not load model: Asynchronous connection has been closed)
>  
> <_InactiveRpcError of RPC that terminated with:
> object-detector-python_1  |     status = StatusCode.UNAVAILABLE
> object-detector-python_1  |     details = "failed to connect to all addresses"
> object-detector-python_1  |     debug_error_string = "{"created":"@1659425957.027178280","description":"Failed to pick subchannel","file":"src/core/ext/filters/client_channel/client_channel.cc","file_line":3217,"referenced_errors":[{"created":"@1659425957.027173640","description":"failed to connect to all addresses","file":"src/core/lib/transport/error_utils.cc","file_line":165,"grpc_status":14}]}"

### **Environment**

- Axis device model: P3265-LVE
- Axis device firmware version: 10.11.76 (also tried 10.10.73 but didn't work)
- SDK VERSION: 1.2.1
- docker daemon with Compose: 1.2.3

Please help me, thanks in advance.

_Originally posted by @JENNSHIUAN in https://github.com/AxisCommunications/acap-computer-vision-sdk-examples/issues/50#issuecomment-1202143368_

**Comment by Corallo** [comment]
Hi @JENNSHIUAN
I converted your question to a new discussion, as it is not related to the problem in that issue.
In your case, It seems that the problem is that the model that you have can't be loaded on the camera.
If you are using yolov4 it is likely that you have compatibility issues, as it is not a model made to run on edge devices.
You could check by SSH into the camera and running ` journalctl -u larod`

**Reply by igor-rud-kr** [reply]
@JENNSHIUAN, did you quantize the model?

I followed the following tutorial: [Retrain SSD MobileNet V1 detector for the Edge TPU (TF1)](https://colab.research.google.com/github/google-coral/tutorials/blob/master/retrain_detection_qat_tf1.ipynb), but I am getting the following error: 

Could not load model: Could not set NNAPI delegate

is there a tutorial on how to re-train SSD MobileNet V2 detector for the Edge TPU?

**Reply by JENNSHIUAN** [reply]
Hi @Corallo 
Thank you for your prompt response.
As you said,  I  do use yolov4.  
I converted the darknet yolov4 to tflite float32 from[ tensorflow-yolov4-tflite](https://github.com/hunglc007/tensorflow-yolov4-tflite) tutorial in tensorflow 2.3.0. 
So can't I run yolov4 on edge devices? or do I need to convert to uint8 or represent it in a smaller bit?
FYI,  I also check by SSH into the camera and running` journalctl -u larod`
The following is what i got:
> -- Journal begins at Tue 2022-08-02 21:42:34 CST, ends at Tue 2022-08-02 23:40:01 CST. --
> Aug 02 23:39:15 axis-b8a44f285981 larod[7322]: Created a new session ID: 4, client: :1.10102
> Aug 02 23:39:21 axis-b8a44f285981 systemd[1]: larod.service: Main process exited, code=killed, status=9/KILL
> Aug 02 23:39:21 axis-b8a44f285981 systemd[1]: larod.service: Failed with result 'signal'.
> Aug 02 23:39:21 axis-b8a44f285981 systemd[1]: larod.service: Consumed 23min 34.103s CPU time.
> Aug 02 23:39:21 axis-b8a44f285981 systemd[1]: larod.service: Scheduled restart job, restart counter is at 4.
> Aug 02 23:39:21 axis-b8a44f285981 systemd[1]: Stopped Machine learning service.
> Aug 02 23:39:21 axis-b8a44f285981 systemd[1]: larod.service: Consumed 23min 34.103s CPU time.
> Aug 02 23:39:21 axis-b8a44f285981 systemd[1]: Starting Machine learning service...
> Aug 02 23:39:22 axis-b8a44f285981 systemd[1]: Started Machine learning service.
> Aug 02 23:39:22 axis-b8a44f285981 larod[55223]: Service started
> Aug 02 23:39:23 axis-b8a44f285981 larod[55223]: Created a new session ID: 0, client: :1.10124

**Reply by JENNSHIUAN** [reply]
@igor-rud-kr 
Hi, I did. I converted my yolov4 model to tflite float32.

**Reply by Corallo** [reply]
@JENNSHIUAN 
Would it be possible to have the tflite version of the model that you are using? You can randomize the weights if needed.

**Reply by JENNSHIUAN** [reply]
Sure, I sent it to your personal email, please check.

**Reply by Corallo** [reply]
Thanks for reporting this issue.
I have forwarded the issue to the right team, we will get back to you when we have updates.
I'm going to move this back to the issues tab, as it is actually an unexpected behavior,

**Reply by Corallo** [reply]
Hi @JENNSHIUAN 

It seems that the model you have is quantized with dynamic range quantization, while it should instead be quantized with integer quantization. See: https://www.tensorflow.org/lite/performance/post_training_quant 
Did you try the int8 version of the quantization from that repository?

**Reply by JENNSHIUAN** [reply]
Sure, I also tried to quantize to uint8 using tensorflow API [tflite converter](https://www.tensorflow.org/lite/convert?hl=zh-tw) with the following tutorial: [Tensorflowlite-Samples](https://github.com/k5iogura/tensorflowlite-samples/blob/master/models_zoo/research/object_detection/README_zoo_to_tflite.md), it can not work, though. I'm frustrated about that. Is there any example model to run yolo on axis camera or should I do [Quantization aware training](https://www.tensorflow.org/model_optimization/guide/quantization/training_example) with it?

**Reply by Corallo** [reply]
Can you share the int8 version of the converted model?
If you run journalctl -u larod when loading that model, do you have the same error in the log?

Unfortunately no, Yolo is not a network optimized for edge devices, and we never made an example with it, we should however at least have a proper error log that show to users what is the problem with the model instead of just crashing.

**Reply by JENNSHIUAN** [reply]
I've already sent an updated one to you, please check. 
The error message is shown:
> ERROR in Inference: Failed to load model train-608_best_uint8_tflite_convert_fromh5.tflite (Could not load model: Could not verify and build model from buffer)

BTW, I'm trying to retrain a new model for ssd mobilenetv2. As @igor-rud-kr mentioned, in order to avoid the same issue, is there a tutorial on how to re-train SSD MobileNet V2 detector?

**Reply by Corallo** [reply]
We have a general [guide](https://github.com/AxisCommunications/acap-native-sdk-examples/tree/main/tensorflow-to-larod-artpec8) about how to retrain models to keep compatible with artpec8 camera, it is not ssd mobilenet specific.
We recommend you to use the [Tensorflow Object Detector API](https://github.com/tensorflow/models/tree/master/research/object_detection) they have a prebuild [ssd mobilenetv2 model 
](https://github.com/tensorflow/models/blob/master/research/object_detection/configs/tf2/ssd_mobilenet_v2_320x320_coco17_tpu-8.config) that you can retrain. If you care about the performance in terms of inference time of your model, take a look at the guide, because artpec8 cameras perform better if you quantize the model properly

**Comment by Corallo** [comment]
@JENNSHIUAN it took a long time for us to finish the investigation, in the end unfortunately we do not support this version of yolo.
The quantization is not done properly.
https://github.com/AxisCommunications/acap-computer-vision-sdk-examples/issues/65

**Comment by Corallo** [comment, marked as answer]
Hi again, 
It is now possible to run Yolov5 on Artpec-8 on devices running Axis OS >= 11.7
See the guide [here](https://github.com/AxisCommunications/axis-model-zoo/blob/main/docs/yolov5-on-artpec8.md)
